<a href="https://colab.research.google.com/github/tsurikow/ml-service-hw1/blob/main/output_model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Домашнее задание №1 (base)**

В этом домашнем задании вам будет необходимо:
*  обучить модель регрессии для предсказания стоимости автомобилей;
* реализовать веб-сервис для применения построенной модели на новых данных

**Максимальная оценка за дз**
> Оценка за домашку = $min(\text{ваш балл}, 11)$

**Мягкий дедлайн: 27 ноября 23:59**

**Жесткий дедлайн: 20 декабря 23:59 (конец модуля)**


**Примечание**

В каждой части оцениваются как код, **так и ответы на вопросы.** Вопросы подсвечены синим цветом.

Если нет одного и/или другого, то часть баллов за соответствующее задание снимается.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import random
import seaborn as sns
import pickle

random.seed(42)
np.random.seed(42)

**Задание 0 (0 баллов).**
Для чего фиксируем сиды в домашках?



---


Для воспроизводимости результатов, например, при разбиении датасета на train и test, или при перемешивании признаков в модели.


---



# **Часть 1 | EDA и визуализация**

Первая часть состоит из классических шагов EDA:

- Базовый EDA и обработка признаков (2.5 балла)
- Визуализации признаков и их анализ (1 балл)

Всего можно набрать 3.5 основных балла и 0.65 бонусных. Бонусные задания выделены как **Дополнительное задание/Бонус**. Вы можете выполнять их, чтобы в случае ошибок в основных задачах всё равно набрать за работу максимум. Кроме того, дополнительные задания позволяют вам углубить знания.

Призываем активно использовать их!

## **Простейший EDA и обработка признаков (2.5 балла)**

In [2]:
df_train = pd.read_csv('https://raw.githubusercontent.com/Murcha1990/MLDS_ML_2022/main/Hometasks/HT1/cars_train.csv')
df_test = pd.read_csv('https://raw.githubusercontent.com/Murcha1990/MLDS_ML_2022/main/Hometasks/HT1/cars_test.csv')

print("Train data shape:", df_train.shape)
print("Test data shape: ", df_test.shape)

Train data shape: (6999, 13)
Test data shape:  (1000, 13)


### **Задание 1 (0.55 балла)**

Выполните операции, направленные на практику основных действий с `pandas`:
- [ ] Отобразите 10 случайных строк тренировочного датасета (0.15 балла)
- [ ] Отобразите первые 5 и последние 5 объектов тестового датасета (0.1 балла)
- [ ] Посмотрите, есть ли в датасете пропуски. Запишите/выведите названия колонок, для которых есть пропущенные значения (0.1 балла)
- [ ] Посмотрите, есть ли в данных явные дубликаты (0.05 балла)


**Бонус**
- [ ] Постройте дашборд, используя [ydata-profilling](https://github.com/ydataai/ydata-profiling)(0.15 балла)

In [3]:
df_train.sample(10)

name  year  selling_price  km_driven  \
6565        Renault KWID Climber 1.0 MT BSIV  2019         300000      35000   
2943                      Maruti Wagon R LXI  2013         225000      58343   
2024                    Hyundai i20 Asta 1.2  2013         360000      30000   
263                     Hyundai i20 1.2 Asta  2010         300000      70000   
4586        Skoda Octavia L and K 1.9 TDI MT  2005         250000     120000   
4479                         Maruti Ciaz ZXi  2016         700000      20000   
4881  Hyundai Grand i10 1.2 Kappa Magna BSIV  2017         445000      27000   
3583          Ford Ecosport 1.5 DV5 MT Trend  2016         515000      68609   
6361                  Hyundai Verna 1.4 VTVT  2014         500000      33400   
4108                     Hyundai i20 Era 1.2  2015         490000      45900   

        fuel       seller_type transmission         owner     mileage  \
6565  Petrol        Individual       Manual   First Owner  23.01 kmpl   
2943  Petrol  Trustmark Dealer       Manual   First Owner  21.79 kmpl   
2024  Petrol        Individual       Manual   First Owner   18.5 kmpl   
263   Petrol        Individual       Manual   First Owner   17.0 kmpl   
4586  Diesel        Individual       Manual   Third Owner   16.4 kmpl   
4479  Petrol        Individual       Manual   First Owner  20.73 kmpl   
4881  Petrol        Individual       Manual   First Owner   18.9 kmpl   
3583  Diesel            Dealer       Manual   First Owner   22.7 kmpl   
6361  Petrol        Individual       Manual   First Owner  17.43 kmpl   
4108  Petrol        Individual       Manual  Second Owner   18.6 kmpl   

       engine  max_power                 torque  seats  
6565   999 CC     67 bhp          91Nm@ 4250rpm    5.0  
2943   998 CC  67.05 bhp          90Nm@ 3500rpm    5.0  
2024  1197 CC  82.85 bhp       113.7Nm@ 4000rpm    5.0  
263   1197 CC     80 bhp  11.4 kgm at 4,000 rpm    5.0  
4586  1896 CC     90 bhp  21.4@ 1,900(kgm@ rpm)    5.0  
4479  1373 CC   91.1 bhp         130Nm@ 4000rpm    5.0  
4881  1197 CC  81.86 bhp      113.75nm@ 4000rpm    5.0  
3583  1498 CC  89.84 bhp    204Nm@ 2000-2750rpm    5.0  
6361  1396 CC  105.5 bhp       135.3Nm@ 5000rpm    5.0  
4108  1197 CC  81.83 bhp       114.7Nm@ 4000rpm    5.0

In [4]:
df_test.head(5)

name  year  selling_price  km_driven  \
0                  Mahindra Xylo E4 BS IV  2010         229999     168000   
1              Tata Nexon 1.5 Revotorq XE  2017         665000      25000   
2                    Honda Civic 1.8 S AT  2007         175000     218463   
3                    Honda City i DTEC VX  2015         635000     173000   
4  Tata Indica Vista Aura 1.2 Safire BSIV  2011         130000      70000   

     fuel seller_type transmission         owner    mileage   engine  \
0  Diesel  Individual       Manual   First Owner  14.0 kmpl  2498 CC   
1  Diesel  Individual       Manual   First Owner  21.5 kmpl  1497 CC   
2  Petrol  Individual    Automatic   First Owner  12.9 kmpl  1799 CC   
3  Diesel  Individual       Manual   First Owner  25.1 kmpl  1498 CC   
4  Petrol  Individual       Manual  Second Owner  16.5 kmpl  1172 CC   

   max_power                   torque  seats  
0    112 bhp  260 Nm at 1800-2200 rpm    7.0  
1  108.5 bhp      260Nm@ 1500-2750rpm    5.0  
2    130 bhp           172Nm@ 4300rpm    5.0  
3   98.6 bhp           200Nm@ 1750rpm    5.0  
4     65 bhp     96  Nm at 3000  rpm     5.0

In [5]:
df_test.tail(5)

name  year  selling_price  km_driven    fuel  \
995            Hyundai i10 Magna 1.1L  2008         250000     100000  Petrol   
996  Hyundai i20 2015-2017 Sportz 1.2  2017         440000      50000  Petrol   
997            Hyundai i20 Era Diesel  2009         340000      40000  Diesel   
998                  Hyundai i10 Asta  2012         350000      25000  Petrol   
999              Honda City i DTec SV  2016         700000     110000  Diesel   

    seller_type transmission         owner     mileage   engine  max_power  \
995  Individual       Manual  Second Owner  19.81 kmpl  1086 CC  68.05 bhp   
996  Individual       Manual  Second Owner   18.6 kmpl  1197 CC  81.83 bhp   
997  Individual       Manual   First Owner   23.0 kmpl  1396 CC     90 bhp   
998  Individual       Manual   First Owner  20.36 kmpl  1197 CC   78.9 bhp   
999  Individual       Manual   First Owner   26.0 kmpl  1498 CC   98.6 bhp   

                       torque  seats  
995          99.04Nm@ 4500rpm    5.0  
996          114.7Nm@ 4000rpm    5.0  
997  22.4 kgm at 1750-2750rpm    5.0  
998          111.8Nm@ 4000rpm    5.0  
999            200Nm@ 1750rpm    5.0

In [6]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6999 entries, 0 to 6998
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           6999 non-null   object 
 1   year           6999 non-null   int64  
 2   selling_price  6999 non-null   int64  
 3   km_driven      6999 non-null   int64  
 4   fuel           6999 non-null   object 
 5   seller_type    6999 non-null   object 
 6   transmission   6999 non-null   object 
 7   owner          6999 non-null   object 
 8   mileage        6797 non-null   object 
 9   engine         6797 non-null   object 
 10  max_power      6803 non-null   object 
 11  torque         6796 non-null   object 
 12  seats          6797 non-null   float64
dtypes: float64(1), int64(3), object(9)
memory usage: 711.0+ KB


In [7]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           1000 non-null   object 
 1   year           1000 non-null   int64  
 2   selling_price  1000 non-null   int64  
 3   km_driven      1000 non-null   int64  
 4   fuel           1000 non-null   object 
 5   seller_type    1000 non-null   object 
 6   transmission   1000 non-null   object 
 7   owner          1000 non-null   object 
 8   mileage        981 non-null    object 
 9   engine         981 non-null    object 
 10  max_power      981 non-null    object 
 11  torque         981 non-null    object 
 12  seats          981 non-null    float64
dtypes: float64(1), int64(3), object(9)
memory usage: 101.7+ KB




---


**В трейне и тесте пропуски есть по колонкам: mileage, engine, max_power, torque, seats**


---



In [8]:
df_train.duplicated().value_counts()

False    6014
True      985
Name: count, dtype: int64

In [9]:
df_test.duplicated().value_counts()

False    938
True      62
Name: count, dtype: int64

In [10]:
!pip install ydata-profiling

  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 390.6/390.6 kB 6.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 296.5/296.5 kB 9.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 686.1/686.1 kB 14.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 104.8/104.8 kB 5.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.5/4.5 MB 25.9 MB/s eta 0:00:00
  Created wheel for htmlmin: filename=htmlmin-0.1.12-py3-none-any.whl size=27081 sha256=20a8bbe37d351b89e7838fe65147ae6be9045dea1db2aab2c450a40c840cd5d5
  Stored in directory: /root/.cache/pip/wheels/dd/91/29/a79cecb328d01739e64017b6fb9a1ab9d8cb1853098ec5966d
Successfully built htmlmin


In [11]:
from ydata_profiling import ProfileReport

In [12]:
profile_train = ProfileReport(df_train, title="Train dashboard")
profile_test = ProfileReport(df_test, title="Test dashboard")

In [13]:
profile_train.to_file("df_train.html")
profile_test.to_file("df_test.html")

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

Мы обнаружили пропуски. Давайте избавимся от них.

- [ ] Заполните пропуски в столбцах медианами. Убедитесь, что после заполнения пропусков не осталось. Заполнение пропусков проводите для обоих наборов данных, если необходимо

**Важно!**

При заполнении пропусков и в тестовом, и тренировочном наборах данных вы определяетесь со стратегией предобработки пропущенных значений при потенциальной работе модели.

Так как в теоретическом случае вы не имеете доступа к тестовой выборке, то заполняемой значение (у нас — медиана) вы считаете про *тренировочному* набору данных и им же заполняете *тестовый*.


---

**Заполнение пропусков будем производить после предобработки данных, поскольку нам сначала необходимо привести признаки к правильному формату**


---







In [14]:
# пропуски заполняем после предобработки

### **Задание 2 (0.5 балла)**

- [ ] Посмотрите, есть ли в трейне объекты с одинаковым признаковым описанием (целевую переменную следует исключить). Если есть, то сколько? (0.1 балла)
- [ ] Отобразите такие объекты (0.15 балла)
- [ ] Удалите повторяющиеся строки. Если при одинаковом признаковом описании цены на автомобили отличаются, то оставьте первую строку по этому автомобилю (0.15 балла)
- [ ]  Обновите индексы строк таким образом, чтобы они шли от 0 без пропусков (0.1 балла)


In [15]:
df_train.duplicated(subset=df_train.columns.difference(['selling_price'])).value_counts()

False    5840
True     1159
Name: count, dtype: int64



---


**1159 объектов с одинаковым признаковым описанием**


---



In [16]:
df_train[df_train.duplicated(subset=df_train.columns.difference(['selling_price']))]

name  year  selling_price  \
254                   Hyundai Grand i10 Sportz  2017         450000   
258                           Maruti Swift VXI  2012         330000   
268                     Maruti Swift Dzire VXI  2014         400000   
324   Jaguar XE 2016-2019 2.0L Diesel Prestige  2017        2625000   
325                              Lexus ES 300h  2019        5150000   
...                                        ...   ...            ...   
6947                    Chevrolet Spark 1.0 LS  2010         140000   
6984  Ford Freestyle Titanium Plus Diesel BSIV  2018         746000   
6989                    Maruti Swift Dzire VDI  2015         625000   
6997                           Tata Indigo CR4  2013         290000   
6998                           Tata Indigo CR4  2013         290000   

      km_driven    fuel seller_type transmission         owner     mileage  \
254       35000  Petrol  Individual       Manual   First Owner   18.9 kmpl   
258       50000  Petrol  Individual       Manual  Second Owner   18.6 kmpl   
268       70000  Petrol  Individual       Manual  Second Owner   19.1 kmpl   
324        9000  Diesel      Dealer    Automatic   First Owner   13.6 kmpl   
325       20000  Petrol      Dealer    Automatic   First Owner  22.37 kmpl   
...         ...     ...         ...          ...           ...         ...   
6947      60000  Petrol  Individual       Manual  Second Owner   18.0 kmpl   
6984      24000  Diesel  Individual       Manual   First Owner   24.4 kmpl   
6989      50000  Diesel  Individual       Manual   First Owner  26.59 kmpl   
6997      25000  Diesel  Individual       Manual   First Owner  23.57 kmpl   
6998      25000  Diesel  Individual       Manual   First Owner  23.57 kmpl   

       engine   max_power               torque  seats  
254   1197 CC      82 bhp       114Nm@ 4000rpm    5.0  
258   1197 CC    85.8 bhp       114Nm@ 4000rpm    5.0  
268   1197 CC    85.8 bhp       114Nm@ 4000rpm    5.0  
324   1999 CC     177 bhp  430Nm@ 1750-2500rpm    5.0  
325   2487 CC  214.56 bhp  202Nm@ 3600-5200rpm    5.0  
...       ...         ...                  ...    ...  
6947   995 CC      62 bhp      90.3Nm@ 4200rpm    5.0  
6984  1498 CC   98.63 bhp  215Nm@ 1750-3000rpm    5.0  
6989  1248 CC      74 bhp       190Nm@ 2000rpm    5.0  
6997  1396 CC      70 bhp  140Nm@ 1800-3000rpm    5.0  
6998  1396 CC      70 bhp  140Nm@ 1800-3000rpm    5.0  

[1159 rows x 13 columns]

In [17]:
df_train = df_train.drop_duplicates(subset=df_train.columns.difference(['selling_price']), keep='first')
df_train = df_train.reset_index(drop=True)

In [18]:
df_train.shape

(5840, 13)

In [19]:
assert df_train.shape == (5840, 13)

Отлично! Мы избавились от маленьких и явных проблем. Теперь перейдем к более сложным недостаткам полученной таблицы.

### **Задание 3 (0.25 балла)**

Вы могли заметить, что с признаками ``mileage, engine, max_power и torque`` всё не очень хорошо. Они распознаются как строки (можно убедиться в этом, вызвав `data.dtypes`). Однако эти переменные не являются категориальными — они — числа. Соответственно, нужно привести их к числовому виду.

**Задача :**
* [ ] Уберите единицы измерения для признаков ``mileage, engine, max_power``.
* [ ] Приведите тип данных к ``float``.
* [ ] Удалите столбец ``torque``


**Важно**
- Все действия нужно производить над обоими датасетами — `train` и `test`.
- Стобец ``torque`` мы удаляем для простоты. В идеальном случае, его также стоило бы предобработать.

In [20]:
df_train = df_train.drop(columns=['torque'])
df_test = df_test.drop(columns=['torque'])

In [21]:
df_train['mileage'] = pd.to_numeric(df_train['mileage'].str.replace(" kmpl", "").str.replace(" km/kg", ""), errors='ignore').astype(float)
df_test['mileage'] = pd.to_numeric(df_test['mileage'].str.replace(" kmpl", "").str.replace(" km/kg", ""), errors='ignore').astype(float)

<ipython-input-21-14d5f96797eb>:1: FutureWarning: errors='ignore' is deprecated and will raise in a future version. Use to_numeric without passing `errors` and catch exceptions explicitly instead
  df_train['mileage'] = pd.to_numeric(df_train['mileage'].str.replace(" kmpl", "").str.replace(" km/kg", ""), errors='ignore').astype(float)
<ipython-input-21-14d5f96797eb>:2: FutureWarning: errors='ignore' is deprecated and will raise in a future version. Use to_numeric without passing `errors` and catch exceptions explicitly instead
  df_test['mileage'] = pd.to_numeric(df_test['mileage'].str.replace(" kmpl", "").str.replace(" km/kg", ""), errors='ignore').astype(float)


In [22]:
df_train['engine'] = pd.to_numeric(df_train['engine'].str.replace(" CC", ""), errors='ignore').astype(float)
df_test['engine'] = pd.to_numeric(df_test['engine'].str.replace(" CC", ""), errors='ignore').astype(float)

<ipython-input-22-2ba721fd83fd>:1: FutureWarning: errors='ignore' is deprecated and will raise in a future version. Use to_numeric without passing `errors` and catch exceptions explicitly instead
  df_train['engine'] = pd.to_numeric(df_train['engine'].str.replace(" CC", ""), errors='ignore').astype(float)
<ipython-input-22-2ba721fd83fd>:2: FutureWarning: errors='ignore' is deprecated and will raise in a future version. Use to_numeric without passing `errors` and catch exceptions explicitly instead
  df_test['engine'] = pd.to_numeric(df_test['engine'].str.replace(" CC", ""), errors='ignore').astype(float)


In [23]:
df_train['max_power'] = pd.to_numeric(df_train['max_power'].str.replace(" bhp", ""), errors='ignore').astype(float)
df_test['max_power'] = pd.to_numeric(df_test['max_power'].str.replace(" bhp", ""), errors='ignore').astype(float)

<ipython-input-23-301ef971e01c>:1: FutureWarning: errors='ignore' is deprecated and will raise in a future version. Use to_numeric without passing `errors` and catch exceptions explicitly instead
  df_train['max_power'] = pd.to_numeric(df_train['max_power'].str.replace(" bhp", ""), errors='ignore').astype(float)
<ipython-input-23-301ef971e01c>:2: FutureWarning: errors='ignore' is deprecated and will raise in a future version. Use to_numeric without passing `errors` and catch exceptions explicitly instead
  df_test['max_power'] = pd.to_numeric(df_test['max_power'].str.replace(" bhp", ""), errors='ignore').astype(float)


**Теперь можно удалить пропуски**

In [24]:
df_train[['mileage','engine','max_power','seats']] = df_train[['mileage','engine','max_power','seats']].fillna(df_train[['mileage','engine','max_power','seats']].median())
df_test[['mileage','engine','max_power','seats']] = df_test[['mileage','engine','max_power','seats']].fillna(df_train[['mileage','engine','max_power','seats']].median())

In [25]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5840 entries, 0 to 5839
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           5840 non-null   object 
 1   year           5840 non-null   int64  
 2   selling_price  5840 non-null   int64  
 3   km_driven      5840 non-null   int64  
 4   fuel           5840 non-null   object 
 5   seller_type    5840 non-null   object 
 6   transmission   5840 non-null   object 
 7   owner          5840 non-null   object 
 8   mileage        5840 non-null   float64
 9   engine         5840 non-null   float64
 10  max_power      5840 non-null   float64
 11  seats          5840 non-null   float64
dtypes: float64(4), int64(3), object(5)
memory usage: 547.6+ KB


In [26]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           1000 non-null   object 
 1   year           1000 non-null   int64  
 2   selling_price  1000 non-null   int64  
 3   km_driven      1000 non-null   int64  
 4   fuel           1000 non-null   object 
 5   seller_type    1000 non-null   object 
 6   transmission   1000 non-null   object 
 7   owner          1000 non-null   object 
 8   mileage        1000 non-null   float64
 9   engine         1000 non-null   float64
 10  max_power      1000 non-null   float64
 11  seats          1000 non-null   float64
dtypes: float64(4), int64(3), object(5)
memory usage: 93.9+ KB


### **Задание 4 (0.1 балла)**

Теперь, когда не осталось пропусков, давайте преобразуем столбцы к более подходящим типам. А именно столбцы ``engnine`` и ``seats`` к приведем к `int`.

- [ ] Осуществите приведение столбцов к необходимому типу.

In [27]:
df_train['engine'] = df_train['engine'].astype(int)
df_train['seats'] = df_train['seats'].astype(int)
df_test['engine'] = df_test['engine'].astype(int)
df_test['seats'] = df_test['seats'].astype(int)

### **Задание 5 (0.1 балла)**

Отлично! Мы провели "косметическую" предобработку и теперь готовы сделать важный шаг в контексте анализа данных. А именно — посмотреть на статистики!

**Ваша задача:**
- [ ] Посчитайте основные статистики по числовым столбцам для трейна и теста
- [ ] Посчитайте основные статистики по категориальным столбцам для трейна и теста

**Подсказка:**

Используте ``.describe()`` с нужным(и) аргументом(-ами).

**Примечание:**

Более корректно рассматривать статистики до заполнения пропусков и после, чтобы убедиться, что мы не внесли каких-либо серьезных сдвигов в изначальные рапсределения.

In [28]:
df_train.describe()

year  selling_price     km_driven      mileage       engine  \
count  5840.000000   5.840000e+03  5.840000e+03  5840.000000  5840.000000   
mean   2013.428425   5.229601e+05  7.395224e+04    19.430841  1429.448973   
std       4.095622   5.354320e+05  6.007114e+04     3.985973   485.662834   
min    1983.000000   2.999900e+04  1.000000e+00     0.000000   624.000000   
25%    2011.000000   2.500000e+05  3.900000e+04    16.950000  1197.000000   
50%    2014.000000   4.050000e+05  7.000000e+04    19.370000  1248.000000   
75%    2017.000000   6.400000e+05  1.000000e+05    22.300000  1498.000000   
max    2020.000000   1.000000e+07  2.360457e+06    42.000000  3604.000000   

         max_power        seats  
count  5840.000000  5840.000000  
mean     87.906371     5.426712  
std      31.640418     0.982344  
min       0.000000     2.000000  
25%      68.000000     5.000000  
50%      81.860000     5.000000  
75%      99.000000     5.000000  
max     400.000000    14.000000

In [29]:
df_test.describe()

year  selling_price      km_driven     mileage      engine  \
count  1000.000000   1.000000e+03    1000.000000  1000.00000  1000.00000   
mean   2013.681000   6.179010e+05   71393.341000    19.33881  1454.87600   
std       4.012149   7.585539e+05   48486.218662     3.95175   521.99574   
min    1995.000000   3.100000e+04    1303.000000     0.00000   624.00000   
25%    2011.000000   2.500000e+05   37000.000000    16.55000  1197.00000   
50%    2014.000000   4.349990e+05   61500.000000    19.37000  1248.00000   
75%    2017.000000   6.700000e+05  100000.000000    22.30000  1582.00000   
max    2020.000000   6.000000e+06  375000.000000    32.26000  3604.00000   

         max_power        seats  
count  1000.000000  1000.000000  
mean     90.841670     5.403000  
std      34.893389     0.912921  
min      34.200000     4.000000  
25%      69.000000     5.000000  
50%      82.425000     5.000000  
75%     102.000000     5.000000  
max     280.000000     9.000000

In [30]:
df_train.describe(include=[object])

name    fuel seller_type transmission        owner
count                     5840    5840        5840         5840         5840
unique                    1924       4           3            2            5
top     Maruti Swift Dzire VDI  Diesel  Individual       Manual  First Owner
freq                        91    3177        5223         5336         3603

In [31]:
df_test.describe(include=[object])

name    fuel seller_type transmission        owner
count                  1000    1000        1000         1000         1000
unique                  621       4           3            2            5
top     Maruti Alto 800 LXI  Diesel  Individual       Manual  First Owner
freq                     15     534         837          877          623

In [32]:
assert df_train.shape == (5840, 12) # изменил к-во столбцов на 12, потому что удален стобец torque

## **Визуализации (1 балл + 0.5 бонус)**



Визуализация данных — важный шаг в работе. Визуализировать данные необходимо, например, чтобы:

- Оценить распределения признаков самих по себе (это может натоклнуть вас на мысли о модели, которую можно использовать)
- Сравнить распределения на `train` и `test` — чтобы проверить, насколько информация, на которой вы будете обучаться согласуется с той, на которой модель должна работать
- Оценить есть ли явная связь признаков с целевой переменной

**Важно:**

Если распределения на `train` и `test` не совпадают, это не значит, что нужно перемешивать данные! Более корректно актуализировать задачу и уточнить, а не устарели ли данные `train`. Также полезным может быть собрать новую тестовую выборку, смешав те, что имеются сейчас.

**Если вы будете подгонять распределения, то можете встретиться с переобучением!**

### **Задание 6 (0.5 балла)**

Шаг 1.
- [ ] Воспользуйтесь `pairplot` из библиотеки `seabron`, чтобы визуализировать попарные распределения числовых признаков для `train`
- [ ] По полученному графику ответьте на вопросы:
 - Можно ли предположить на основе распределений связь признаков с целевой переменной?
 - Можно ли предположить на основе распределений выдвинуть гипотезу о корреляциях признаков?

Шаг 2.

- [ ] Постройте pairplot по тестовым данным
- [ ] Ответьте на вопрос "Похожими ли оказались совокупности при разделении на трейн и тест?"



---


При визуальном анализе распределений мы можем обнаружить следующие связи признаков с целевой переменной:

1.   Год выпуска: чем свежее год выпуска, тем вероятнее более высокое значение целевой переменной
2.   Максимальная мощность: чем выше максимальная мощность, тем вероятнее более высокое значение целевой переменной
3.   Объем двигателя: чем выше объем, тем вероятнее более высокое значение целевой переменной



При визуальном анализе распределений мы можем выдвинуть следующие гипотезы о корреляциях признаков:

1.   Распределения максимальной мощности и объема двигателя могут говорить о сильной положительной линейной зависимости между этими признаками
2.   Распределения объема двигателя и расхода топлива могут говорить о достаточно существенной корреляции, похожим образом по отношению к расходу топлива ведет себя максимальная мощность двигателя
3.   Распределения количества сидений и расхода тоже могут говорить о существенной корреляции между этими признаками

Для подтверждения гипотез, безусловно, требуются дополнительные исследования

После разделения на тренировочную и тестовую выборки, распределения значений остались в достаточной степени похожими, что видно по совмещенным попарно графикам. Так же можно предположить, что в обучающей выборке присутствуют выбросы, которые в тестовых данных не наблюдаются, поэтому при оценивании распределений между двумя выборками важно сохранять диапазон значений по осям.


---



### **Задание 7 (0.5 балла)**

И так, вы выдвинули гипотезы о наличии связи. Теперь давайте оценим эту связь в числах.

**Задание:**
- [ ] Получите значения коэффициента корреляции Пирсона для тренировочного набора данных при помощи `pd.corr()`
- [ ] По полученным корреляциям постройте тепловую карту (`heatmap` из бибилотеки seaborn)


- [ ] Ответьте на вопросы:
 - Какие 2 признака наименее скоррелированы между собой?
 - Между какими наблюдается довольно сильная положительная линейная зависимость?
 - Правильно ли, опираясь на данные, утверждать, что чем меньше год, тем, скорее всего, больше километров проехала машина к дате продажи?



---


Наименее скоррелированы между собой год выпуска/объем двигателя и пробег/максимальная мощность.

Сильная положительная линейная зависимость наблюдается между признаками: количество сидений/объем двигателя и объем двигателя/максимальная мощность.

Данные показывают, что чем свежее год выпуска, тем выше вероятность, что автомобиль проехал меньше километров. Но это скорее не утверждение, а гипотеза, которая требует дополнительного анализа.


---





### **Бонус (0.5 балла)**

Если вам кажется, что мы не попросили вас нарисовать какие-то очень важные зависимости, нарисуйте их **и поясните.**




---
Ящики с усами показывают нам выбросы по правилу крайних квартилей с 1.5 среднеквартильным размахом \
В идеале нам стоило бы изучить эти выбросы на предмет того, не являются ли они ошибками в данных.

---



# **Часть 2 | Модель только на вещественных признаках**

В этой части вам предстоит обучить модель только на вещественных признаках. Почему только на них?

Чем больше признаковое пространство — чем сложнее модель. А чем модель проще — тем лучше для скорости работы и интерпретации признаков.

За задания этой части вы можете набрать 1.25 балла;

### **Задание 8 (0.05 балла)**

Разбейте данные на тренировочный и тестовый наборы. Перед разбиением создайте копию датафрейма, который будет хранить только вещественные признаки и используйте его (то есть категориальные столбцы (все, кроме seats) необходимо удалить).

В переменные y_train и y_test запишите значения целевых переменных.

In [33]:
y_train = df_train['selling_price'].copy()
X_train = df_train[['year', 'km_driven', 'mileage', 'engine', 'max_power', 'seats']].copy()

In [34]:
assert X_train.shape == (5840, 6)

In [35]:
y_test = df_test['selling_price'].copy()
X_test = df_test[['year', 'km_driven', 'mileage', 'engine', 'max_power', 'seats']].copy()

In [36]:
assert X_test.shape == (1000, 6)

### **Задание 9 (0.2 балла)**

Построим нашу первую модель!
- [ ] Обучите классическую линейную регрессию с дефолтными параметрами. Посчтитайте $R^2$ и $MSE$ для трейна и для теста.
- [ ] Сделайте выводы по значениям метрик качества.

**Примечание:**

Здесь и далее $R^2$ и $MSE$ для трейна и для теста выводите везде, где требуется обучать модели, даже если в явном виде этого не просят. Иначе непонятно, как понять, насколько успешны наши эксперименты.

In [37]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, root_mean_squared_error as RMSE, mean_squared_error as MSE

regr = LinearRegression()

regr.fit(X_train, y_train)

y_pred = regr.predict(X_test)
y_train_pred = regr.predict(X_train)

print("___test___")
print("MSE: %.2f" % MSE(y_test, y_pred))
print("RMSE: %.2f" % RMSE(y_test, y_pred))
print("R^2: %.2f" % r2_score(y_test, y_pred))
print("___train___")
print("MSE: %.2f" % MSE(y_train, y_train_pred))
print("RMSE: %.2f" % RMSE(y_train, y_train_pred))
print("R^2: %.2f" % r2_score(y_train, y_train_pred))

___test___
MSE: 233298779730.45
RMSE: 483010.12
R^2: 0.59
___train___
MSE: 116874153930.03
RMSE: 341868.62
R^2: 0.59




---


Метрики качества говорят о том, что у модели получается предсказывать целевую переменную, но ошибается она достаточно сильно и часто, поэтому нужно работать над улучшением модели, чтобы добиться более приемлемых результатов.
\
Так же можно предположить, что мы столкнулись с переобучением, на основании того, что ошибка на тесте сильно выше ошибки на трейне.


---



### **Задание 10 (0.15 балла)**

Всегда есть место совершенству. Поэтому давайте попробуем улучшить модель. При помощи стандартизации признаков.

- [ ] Стандартизируйте значения в тренировочных и тестовых данных. Стандартизатор **обучайте только на `train`**.

In [38]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

scaler.fit(X_train)

X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

regr = LinearRegression()

regr.fit(X_train, y_train)

y_pred = regr.predict(X_test)
y_train_pred = regr.predict(X_train)

print("___test___")
print("MSE: %.2f" % MSE(y_test, y_pred))
print("RMSE: %.2f" % RMSE(y_test, y_pred))
print("R^2: %.2f" % r2_score(y_test, y_pred))
print("___train___")
print("MSE: %.2f" % MSE(y_train, y_train_pred))
print("RMSE: %.2f" % RMSE(y_train, y_train_pred))
print("R^2: %.2f" % r2_score(y_train, y_train_pred))


___test___
MSE: 233298779730.45
RMSE: 483010.12
R^2: 0.59
___train___
MSE: 116874153930.03
RMSE: 341868.62
R^2: 0.59


### **Задание 11 (0.1 балла)**

Хотя стандартизация не помогла сильно прибавить в качестве она открыла возможность интерпретировать важность признаков в модели. Правило интерпретации такое:

Чем больше коэффициент $\beta_i$ по модулю, тем важнее признак.

**Ответьте на вопрос:**

- [ ] Какой признак оказался наиболее информативным в предсказании цены?

In [39]:
scaled_X_train = pd.DataFrame(X_train)
scaled_X_train.apply(lambda x: abs(x)).mean()

0    0.794951
1    0.617992
2    0.782141
3    0.737595
4    0.704728
5    0.736184
dtype: float64



---


Самым информативным в предсказании цены оказался год выпуска.


---



### **Задание 12 (0.25 балла)**

Попробуем улучшить нашу модель с помощью применения регуляризации. Для этого воспльзуемся `Lasso` регрессией.  Кроме того, попробуйте использовать её теоретическое свойство отбора признаков, за счет зануления незначимых коэффициентов.

**Задание:**

- [ ] Обучите Lasso регрессию на тренировочном наборе данных с нормализованными признаками. Оцените её качество
- [ ] Проверьте, занулила ли L1-регуляризация с параметрами по умолчанию какие-нибудь веса? Предположите почему.

In [40]:
from sklearn.linear_model import Lasso

lassoreg = Lasso()

lassoreg.fit(X_train, y_train)

y_pred_lasso = lassoreg.predict(X_test)
y_train_pred_lasso = lassoreg.predict(X_train)

print("___test___")
print("MSE: %.2f" % MSE(y_test, y_pred_lasso))
print("RMSE: %.2f" % RMSE(y_test, y_pred_lasso))
print("R^2: %.2f" % r2_score(y_test, y_pred_lasso))
print("___train___")
print("MSE: %.2f" % MSE(y_train, y_train_pred_lasso))
print("RMSE: %.2f" % RMSE(y_train, y_train_pred_lasso))
print("R^2: %.2f" % r2_score(y_train, y_train_pred_lasso))

___test___
MSE: 233299450599.17
RMSE: 483010.82
R^2: 0.59
___train___
MSE: 116874153940.14
RMSE: 341868.62
R^2: 0.59




---


Качество модели с параметрами по-умолчанию соответствует модели обычной линейной регрессии.


---



In [41]:
print(lassoreg.intercept_)
print(lassoreg.sparse_coef_)

522960.0910958868
  (0, 0)	152772.39021419562
  (0, 1)	-44157.689857942605
  (0, 2)	28832.04339588388
  (0, 3)	61173.66240922125
  (0, 4)	322808.23691946967
  (0, 5)	-23206.25564908013




---


Зануленных коэффициентов нет с параметрами по-умолчанию


---



### **Задание 13 Финальный рывок (0.5 балла)**

До этого мы с вами использовали `train` для обучения и `test` для прогнозирования. Но у нас есть ещё одна задача — подобрать оптимальные параметры модели. Для этого используем кросс-валидацию, описанную на семинарах.

Кроме того, выжмем максимум из модификаций регрессии. Построим `ElasticNet`. И сделаем всё по порядку.

**Ваша задача 1:**

- [ ] Перебором по сетке (c 10-ю фолдами) подберите оптимальные параметры для Lasso-регрессии. Вам пригодится класс [GridSearchCV](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html).
- [ ] Ответьте на вопросы:
 - Сколько грид-сёрчу пришлось обучать моделей?
 - Какой коэффициент регуляризации у лучшей из перебранных моделей? Занулились ли какие-нибудь из весов при такой регуляризации?

In [42]:
from sklearn.model_selection import GridSearchCV

param_grid = {
    'alpha': np.arange(1, 1500, 10)
}

grid_search = GridSearchCV(estimator=Lasso(),
                           param_grid=param_grid,
                           cv=10,
                           scoring='neg_mean_squared_error')

grid_search.fit(X_train, y_train)

y_pred_lasso = grid_search.predict(X_test)
y_train_pred_lasso = grid_search.predict(X_train)

print(f"Best score: {grid_search.best_score_:.3f}")
print(f"Best parameters: {grid_search.best_params_}")

Best score: -121303403782.183
Best parameters: {'alpha': 1131}





---
Коэффициент регуляризации у лучшей модели: 1131 \
До этого предварительно перебирались значения из разных диапазонов, коэффициент всегда стремился к ~1100-1200

---



In [43]:
print("___test___")
print("MSE: %.2f" % MSE(y_test, y_pred_lasso))
print("RMSE: %.2f" % RMSE(y_test, y_pred_lasso))
print("R^2: %.2f" % r2_score(y_test, y_pred_lasso))
print("___train___")
print("MSE: %.2f" % MSE(y_train, y_train_pred_lasso))
print("RMSE: %.2f" % RMSE(y_train, y_train_pred_lasso))
print("R^2: %.2f" % r2_score(y_train, y_train_pred_lasso))

___test___
MSE: 234068066758.98
RMSE: 483805.82
R^2: 0.59
___train___
MSE: 116886766629.78
RMSE: 341887.07
R^2: 0.59


In [44]:
print(grid_search.best_estimator_.intercept_)
print(grid_search.best_estimator_.sparse_coef_)

522960.0910958868
  (0, 0)	152530.66784097048
  (0, 1)	-42906.17683918457
  (0, 2)	26768.35945901569
  (0, 3)	55811.5475542441
  (0, 4)	323779.5727109988
  (0, 5)	-19977.93804421957


In [45]:
from sklearn.model_selection import ParameterGrid

pg = ParameterGrid(param_grid)

print(f"Number of models: {len(pg)*10}")

Number of models: 1500




---
В данном случае мы обучили 1500 моделей


---



**Ваша задача 2:**

- [ ] Перебором по сетке (c 10-ю фолдами) подберите оптимальные параметры для [ElasticNet](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.ElasticNet.html) регрессии.
- [ ] Ответьте на вопрос:
 - Сколько грид-сёрчу пришлось обучать моделей?
 - Какие гиперпараметры соответствуют лучшей (по выбранной метрике качества) из перебранных моделей?

In [46]:
from sklearn.linear_model import ElasticNet

param_grid = {
              'alpha': [0.1,1,10,0.01],
              'l1_ratio': np.arange(0.40,1.00,0.03),
              'tol': [0.0001,0.001]
             }

grid_search = GridSearchCV(estimator=ElasticNet(),
                           param_grid=param_grid,
                           cv=10,
                           scoring='neg_mean_squared_error')

grid_search.fit(X_train, y_train)

y_pred_enet = grid_search.predict(X_test)
y_train_pred_enet = grid_search.predict(X_train)

print("Best parameters : {}".format(grid_search.best_params_))
print("Best cross validation score: {:.2f}".format(grid_search.best_score_))

Best parameters : {'alpha': 0.1, 'l1_ratio': 0.8500000000000004, 'tol': 0.001}
Best cross validation score: -121286855978.82


In [47]:
print("___test___")
print("MSE: %.2f" % MSE(y_test, y_pred_enet))
print("RMSE: %.2f" % RMSE(y_test, y_pred_enet))
print("R^2: %.2f" % r2_score(y_test, y_pred_enet))
print("___train___")
print("MSE: %.2f" % MSE(y_train, y_train_pred_enet))
print("RMSE: %.2f" % RMSE(y_train, y_train_pred_enet))
print("R^2: %.2f" % r2_score(y_train, y_train_pred_enet))

___test___
MSE: 235195411454.59
RMSE: 484969.50
R^2: 0.59
___train___
MSE: 116908142911.16
RMSE: 341918.33
R^2: 0.59


In [48]:
from sklearn.model_selection import ParameterGrid

pg = ParameterGrid(param_grid)

print(f"Number of models: {len(pg)*10}")

Number of models: 1600




---

Было обучено 1600 моделей \
Лучшими коэффициентами оказались: 'alpha': 0.1, 'l1_ratio': 0.8500000000000004, 'tol': 0.001

---



# **Часть 3| Добавляем категориальные фичи**

Попробуем для улучшения модели дать ей больше признаков. Добавим категориальные фичи.

За эту часть можно набрать 0.75 основных балла и 0.25 бонусных.


### **Задание 14 (0.1 балла)** Проанализируйте столбец `name`. Очевидно, что эта переменная является категориальной, однако категорий в ней много.

В этом домашнем задании мы предлагаем удалить его.

**Ваша задача:**
- [ ] Удалить столбец`name`

In [49]:
# не будем удалять столбец, закодируем с помощью TargetEncoder
df_train['name'].nunique()

1924

In [50]:
df_train['name'].value_counts().head(20)

name
Maruti Swift Dzire VDI           91
Maruti Alto 800 LXI              58
Maruti Alto LXi                  56
Maruti Swift VDI                 48
Maruti Swift VDI BSIV            47
Maruti Alto K10 VXI              39
Maruti Ertiga VDI                37
Hyundai EON Era Plus             36
Maruti Alto LX                   34
Maruti Swift Dzire VXI           34
Maruti Ritz VDi                  33
Maruti Wagon R VXI BS IV         33
Maruti 800 AC                    30
Ford Figo Diesel Titanium        25
Maruti Swift Dzire VDi           25
Mahindra XUV500 W6 2WD           23
Renault KWID RXT                 23
Hyundai Santro Xing GLS          22
Hyundai EON Magna Plus           22
Maruti Alto K10 2010-2014 VXI    22
Name: count, dtype: int64

In [51]:
from sklearn.preprocessing import TargetEncoder

X_train_cat = df_train.drop(columns=['selling_price']).copy()

targetenc = TargetEncoder(categories='auto', target_type='continuous', smooth='auto')
X_train_cat['name'] = targetenc.fit_transform(X_train_cat[['name']], y_train)

X_test_cat = df_test.drop(columns=['selling_price']).copy()
X_test_cat['name'] = targetenc.transform(X_test_cat[['name']])

In [52]:
X_train_cat.head()

name  year  km_driven    fuel seller_type transmission  \
0  584664.475473  2014     145500  Diesel  Individual       Manual   
1  624960.574087  2014     120000  Diesel  Individual       Manual   
2  523661.040668  2010     127000  Diesel  Individual       Manual   
3  170000.000000  2007     120000  Petrol  Individual       Manual   
4  500000.000000  2017      45000  Petrol  Individual       Manual   

          owner  mileage  engine  max_power  seats  
0   First Owner    23.40    1248      74.00      5  
1  Second Owner    21.14    1498     103.52      5  
2   First Owner    23.00    1396      90.00      5  
3   First Owner    16.10    1298      88.20      5  
4   First Owner    20.14    1197      81.86      5

In [53]:
X_test_cat.head()

name  year  km_driven    fuel seller_type transmission  \
0  245000.000000  2010     168000  Diesel  Individual       Manual   
1  522960.091096  2017      25000  Diesel  Individual       Manual   
2  522960.091096  2007     218463  Petrol  Individual    Automatic   
3  698660.989991  2015     173000  Diesel  Individual       Manual   
4  522960.091096  2011      70000  Petrol  Individual       Manual   

          owner  mileage  engine  max_power  seats  
0   First Owner     14.0    2498      112.0      7  
1   First Owner     21.5    1497      108.5      5  
2   First Owner     12.9    1799      130.0      5  
3   First Owner     25.1    1498       98.6      5  
4  Second Owner     16.5    1172       65.0      5

В другом случае, конечно, мы могли бы предобработать данный столбец. В качестве бонуса предлагаем вам придумать и реализовать алгоритм предобработки.

### **Бонус 0.5 балла**
- [ ] Предобработайте столбец `name`, чтобы избежать его удаления



---


Столбец name был закодирован TargetEncoder'ом по целевой переменной. Удалять данный столбец нельзя, поскольку в нем, возможно, содержится самая полезная информация для прогнозирования целевой переменной, а именно: производитель, модель и комплектация авто. Мы можем столкнуться с переобучением, но в TargetEncoder есть smoothing и кросс-валидация, которые должны нас от этого уберечь.

---



In [54]:
assert X_train_cat.shape == (5840, 11) # изменено с 10 на 11, потому что столбец name не удалялся

In [55]:
X_train_cat.describe(include='object')

fuel seller_type transmission        owner
count     5840        5840         5840         5840
unique       4           3            2            5
top     Diesel  Individual       Manual  First Owner
freq      3177        5223         5336         3603

### **Задание 15 (0.4 балла)**

- [ ] Закодируйте категориалльные фичи и ``seats`` методом OneHot-кодирования. Обратите внимание, что во избежание мультиколлинеарности следует избавиться от одного из полученных столбцов при кодировании каждого признака методом OneHot.

In [56]:
# перед кодировкой категориальных фич, изучим значения этих столбцов, возможно в них есть ошибки

X_train_cat['fuel'].value_counts()

fuel
Diesel    3177
Petrol    2579
CNG         50
LPG         34
Name: count, dtype: int64

In [57]:
X_train_cat['seller_type'].value_counts()

seller_type
Individual          5223
Dealer               592
Trustmark Dealer      25
Name: count, dtype: int64

In [58]:
X_train_cat['transmission'].value_counts()

transmission
Manual       5336
Automatic     504
Name: count, dtype: int64

In [59]:
X_train_cat['owner'].value_counts()

owner
First Owner             3603
Second Owner            1639
Third Owner              454
Fourth & Above Owner     140
Test Drive Car             4
Name: count, dtype: int64



---


В целом ошибки не обнаружены, но смущает наличие редко встречающихся, и, возможно, не очень полезных категорий:


1.   Trustmark Dealer в seller_type
2.   Test Drive Car в owner

В данном задании с этим ничего делать не будем.


---




In [60]:
X_train_cat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5840 entries, 0 to 5839
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   name          5840 non-null   float64
 1   year          5840 non-null   int64  
 2   km_driven     5840 non-null   int64  
 3   fuel          5840 non-null   object 
 4   seller_type   5840 non-null   object 
 5   transmission  5840 non-null   object 
 6   owner         5840 non-null   object 
 7   mileage       5840 non-null   float64
 8   engine        5840 non-null   int64  
 9   max_power     5840 non-null   float64
 10  seats         5840 non-null   int64  
dtypes: float64(3), int64(4), object(4)
memory usage: 502.0+ KB


In [61]:
from sklearn.preprocessing import OneHotEncoder # или можно использовать get_dummies из библиотеки pandas

X_train_cat['seats'] = X_train_cat['seats'].astype('object')
X_test_cat['seats'] = X_test_cat['seats'].astype('object')

categorical_columns = X_train_cat.select_dtypes(include=['object']).columns.tolist()

oheenc = OneHotEncoder(sparse_output=False, drop='first', handle_unknown='ignore')

one_hot_encoded = oheenc.fit_transform(X_train_cat[categorical_columns])

one_hot_df = pd.DataFrame(one_hot_encoded, columns=oheenc.get_feature_names_out(categorical_columns))

df_encoded = pd.concat([X_train_cat, one_hot_df], axis=1)

X_train_cat = df_encoded.drop(categorical_columns, axis=1)

In [62]:
X_train_cat.shape

(5840, 24)

In [63]:
# повторим для теста
one_hot_encoded = oheenc.transform(X_test_cat[categorical_columns])

one_hot_df = pd.DataFrame(one_hot_encoded, columns=oheenc.get_feature_names_out(categorical_columns))

df_encoded = pd.concat([X_test_cat, one_hot_df], axis=1)

X_test_cat = df_encoded.drop(categorical_columns, axis=1)

In [64]:
X_train_cat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5840 entries, 0 to 5839
Data columns (total 24 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   name                          5840 non-null   float64
 1   year                          5840 non-null   int64  
 2   km_driven                     5840 non-null   int64  
 3   mileage                       5840 non-null   float64
 4   engine                        5840 non-null   int64  
 5   max_power                     5840 non-null   float64
 6   fuel_Diesel                   5840 non-null   float64
 7   fuel_LPG                      5840 non-null   float64
 8   fuel_Petrol                   5840 non-null   float64
 9   seller_type_Individual        5840 non-null   float64
 10  seller_type_Trustmark Dealer  5840 non-null   float64
 11  transmission_Manual           5840 non-null   float64
 12  owner_Fourth & Above Owner    5840 non-null   float64
 13  own

In [65]:
X_test_cat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 24 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   name                          1000 non-null   float64
 1   year                          1000 non-null   int64  
 2   km_driven                     1000 non-null   int64  
 3   mileage                       1000 non-null   float64
 4   engine                        1000 non-null   int64  
 5   max_power                     1000 non-null   float64
 6   fuel_Diesel                   1000 non-null   float64
 7   fuel_LPG                      1000 non-null   float64
 8   fuel_Petrol                   1000 non-null   float64
 9   seller_type_Individual        1000 non-null   float64
 10  seller_type_Trustmark Dealer  1000 non-null   float64
 11  transmission_Manual           1000 non-null   float64
 12  owner_Fourth & Above Owner    1000 non-null   float64
 13  owne

### **Задание 16 (0.25 балла)**
Повторим то, что делали на прошлом шаге для моделей на вещественных признаках, однако теперь с моделью `Ridge`.


**Ваша задача:**
- [ ] Переберите параметр регуляризации `alpha` для гребневой (ridge) регрессии с помощью класса `GridSearchCV` В качестве параметров при объявлении GridSearchCV кроме модели укажите метрику качества $R^2$. Кроссвалидируйтесь по 10-ти фолдам.
- [ ] Ответье на вопрос: Удалось ли улучшить качество прогнозов?

In [66]:
# задание понял так, что нужно обучить только Ridge,
# потому что четкой задачи на переобучение остальных моделей не вижу
from sklearn.linear_model import Ridge
from sklearn.model_selection import GridSearchCV

param_grid = {'alpha': [25,10,4,2,1.0,0.8,0.5,0.3,0.2,0.1,0.05,0.02,0.01]}

grid_search = GridSearchCV(estimator=Ridge(),
                           param_grid=param_grid,
                           cv=10,
                           scoring='r2')

grid_search.fit(X_train_cat, y_train)

y_pred_ridge = grid_search.predict(X_test_cat)
y_train_pred_ridge = grid_search.predict(X_train_cat)

print("Best parameters : {}".format(grid_search.best_params_))
print("Best cross validation score: {:.2f}".format(grid_search.best_score_))

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_ridge.py:216: LinAlgWarning: Ill-conditioned matrix (rcond=6.33722e-17): result may not be accurate.
  return linalg.solve(A, Xy, assume_a="pos", overwrite_a=True).T
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_ridge.py:216: LinAlgWarning: Ill-conditioned matrix (rcond=2.53489e-17): result may not be accurate.
  return linalg.solve(A, Xy, assume_a="pos", overwrite_a=True).T


Best parameters : {'alpha': 0.01}
Best cross validation score: 0.70


/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_ridge.py:216: LinAlgWarning: Ill-conditioned matrix (rcond=1.26744e-17): result may not be accurate.
  return linalg.solve(A, Xy, assume_a="pos", overwrite_a=True).T


In [67]:
print("___test___")
print("MSE: %.2f" % MSE(y_test, y_pred_ridge))
print("RMSE: %.2f" % RMSE(y_test, y_pred_ridge))
print("R^2: %.2f" % r2_score(y_test, y_pred_ridge))
print("___train___")
print("MSE: %.2f" % MSE(y_train, y_train_pred_ridge))
print("RMSE: %.2f" % RMSE(y_train, y_train_pred_ridge))
print("R^2: %.2f" % r2_score(y_train, y_train_pred_ridge))

___test___
MSE: 77004231939.74
RMSE: 277496.36
R^2: 0.87
___train___
MSE: 77938740378.56
RMSE: 279175.11
R^2: 0.73


In [68]:
print(grid_search.best_estimator_.intercept_)
print(grid_search.best_estimator_.coef_)

-51218917.511293
[ 4.62402230e-01  2.54544856e+04 -4.45706789e-01  5.62144457e+03
  4.11479444e+01  6.75302923e+03  3.44211120e+04  1.40901856e+05
 -1.98734565e+03 -7.73748161e+04 -1.26087150e+05 -2.48902775e+05
 -3.44373784e+04 -3.91995194e+04  1.70278236e+06 -3.20693262e+04
  2.55108354e+05 -1.78772977e+05 -2.64820211e+05 -1.96811844e+05
 -1.58257546e+05 -1.45299113e+05 -1.84100632e+05 -3.83657409e+05]


In [69]:
from sklearn.model_selection import ParameterGrid

pg = ParameterGrid(param_grid)

print(f"Number of models: {len(pg)*10}")

Number of models: 130


In [70]:
ridgemodel = grid_search.best_estimator_

In [71]:
file = open('models.pkl','wb')
pickle.dump(targetenc, file)
pickle.dump(oheenc, file)
pickle.dump(ridgemodel, file)
file.close()



---


Без закодированной TargetEncoder переменной name улучшение результата не настолько глобальное, что вполне ожидаемо. Разница в ошибках по трейну и тесту без закодированной фичи name достаточно велика и соответствует первым пробам без категиориальных фич.


---



# **Часть 4. | Бизнесовая (0.5 балла)**

### **Задание 17 (0.5 балла)**

В мире бизнеса очень важно давать оценку качества модели понятную бизнесу, поэтому иногда заказчики приходят с кастомными метриками. Попробуем сделать такую для нашей задачи.

**Описание метрики:**

Среди всех предсказанных цен на авто нужно посчитать долю прогнозов, отличающихся от реальных цен на эти авто не более чем на 10% (в одну или другую сторону)

**Ваша задача:**

- [ ] Реализуйте метрику `business_metric`
- [ ] Посчитайте метрику для всех обученных моделей и определеите, какаю лучше всего решает задачу бизнеса

In [72]:
def business_metric(y_true, y_pred):
    y_true = pd.DataFrame(y_true)
    y_pred = pd.Series(y_pred)
    df_metric = pd.concat([y_true, y_pred], axis=1)
    df_metric.columns = ['true', 'pred']
    # четко не указано, от какой величины считать процент, поэтому взят % от true
    df_metric['metric'] = df_metric.apply(lambda x: 1 if 0.9 <= x['pred']/x['true'] <= 1.1 else 0, axis=1)
    return df_metric['metric'].sum()/df_metric.shape[0]


print("___Business metric___")
print("LinReg: %.3f" % business_metric(y_test, y_pred))
print("Lasso: %.3f" % business_metric(y_test, y_pred_lasso))
print("ENet: %.3f" % business_metric(y_test, y_pred_enet))
print("Ridge (+name): %.3f" % business_metric(y_test, y_pred_ridge))

___Business metric___
LinReg: 0.227
Lasso: 0.224
ENet: 0.222
Ridge (+name): 0.268




---
Лучше всего с задачей справилась модель Ridge с закодированным TargetEncoder'ом стобцом name и параметром 'alpha': 0.02
\
Эта же модель показала лучшие метрики MSE и R2


---



# **Часть 5 (3 балла) | Реализация сервиса на FastAPI**

### **Задание 18 (3 балла)**


Cделайте с помощью FastAPI сервис, который с точки зрения пользователя реализует две функции:

1. на вход в формате json подаются признаки одного объекта, на выходе сервис выдает предсказанную стоимость машины
2. на вход подается csv-файл с признаками тестовых объектов, на выходе получаем файл с +1 столбцом - предсказаниями на этих объектах

С точки зрения реализации это означает следующее:
- средствами pydantic должен быть описан класс базового объекта
- класс с коллецией объектов
- метод post, который получает на вход один объект описанного класса
- метод post, который получает на вход коллекцию объектов описанного класса

Шаблон для сервисной части дан ниже. Код необходимо дополнить и оформить в виде отдельного .py-файла.

Протестируйте сервис на корректность работы и приложите скриншоты (см. ниже).

# **Часть 6 (1 балл) | Оформление результатов**

### **Задание 19 (1 балл)**


**Результаты вашей работы** необходимо разместить в своем Гитхабе. Под результатами понимаем следующее:
* ``.ipynb``-ноутбук со всеми проведёнными вами экспериментами (output'ы ячеек, разумеется, сохранить)
* сохраненный дашборд в любом формате
* ``.py``-файл с реализацией сервиса
* ``.pickle``-файл с сохранёнными весами модели, коэффициентами скейлинга и прочими числовыми значениями, которые могут понадобиться для инференса
* ``.md``-файл с выводами про проделанной вами работе:
    * что было сделано
    * с какими результатами
    * что дало наибольший буст в качестве
    * что сделать не вышло и почему (это нормально, даже хорошо😀)

**За что могут быть сняты баллы в этом пункте:**
* за отсутствие ``.pickle``-файла с весами использованной модели
* за недостаточную аналитику в ``.md``-файле
* за оформление и логику кода (в определённом смысле это тоже элемент оформления решения)

**Как будет выглядет проверка всего домашнего задания?**
1. Ассистент проходит по ссылке на (**открытый**) репозиторий из Энитаска
2. Смотрит ``readme.md``:
    * пожалуйста, приложите в него же скрины работы вашего сервиса -- собирать ваши проекты довольно времязатратно, но хочется убедиться, что всё работает
    * можете в md-файл приложить ссылку на screencast с демонстрацией
3. Просматривает ноутбук с DS-частью
4. Заглядывает в код сервиса
5. Хвалит

# **Часть Благодарственная**

Надеемся, вы честно проделали все пункты, а не просто пролистали досюда. Потому что здесь награда за старания. Пожалуйста, не стоит награждать себя до того, как закончите работать над домашкой!

<details>
<summary><b>Что-то приятное</b></summary>

**Напоминаем, что нашем курсе действует система кото-бонусов** 🐈

На фото по ссылке — сэр кот кого-то из команды курса (преподаватель, помощник преподавателя, ассистенты).

Предлагаем вам угадать — чей это товарищ!

[Первый кот](https://ibb.co/XbnpCTg)

</details>

